Update to Data modeling index page (#430)
neo4j/docs-getting-started: modules/ROOT/pages/data-modeling/index.adoc

The next step is to add <<labels>> and determine <<properties>> for the nodes and relationships.
This helps differentiating entities and thus making it easier to retrieve them from the database:

.Matrix - add labels and properties
image::matrix-whiteboard-model3.svg[500,500,role="popup-link"]

To visualize that in Neo4j Browser or in Query, you can retrieve the information from the xref:appendix/example-data.adoc[*Movies* example dataset] or write the data from scratch:

[source,cypher]
--
CREATE (p1:Person:Actor {name: "Tom Hanks"})-[:ACTED_IN]->(m1:Movie {title: "Cloud Atlas"})<-[:ACTED_IN]-(p2:Person:Actor {name: "Hugo Weaving"}),
(p2)-[:ACTED_IN]->(m2:Movie {title: "The Matrix"})<-[:DIRECTED]-(p3:Person:Director {name: "Lana Wachowski"}),
(p3)-[:DIRECTED]->(m1)
--

To retrieve this information from the database, write the following query:

[source,cypher]
--
MATCH p=()-[:ACTED_IN]->()<-[:DIRECTED]-()
RETURN p
--

The visualization should look like this:

.Matrix - final model in Neo4j 
image::matrix-whiteboard-model4.svg[400,400,role="popup-link"]

[TIP]
====
If you are new to these concepts, refer to xref:cypher.adoc[What is Cypher] for an introduction.
====

> === Domain definition

Author: Why description instead of definition?

> It is important to understand the domain of your dataset when designing data models.

Reviewer: ```suggestion
It is easier to understand the process of designing a graph data model if you consider the domain.
```

As an example, consider the following scenario:

* Two _people_, *Sally* and *John*, +++<u>are friends</u>+++.
* Both *John* and *Sally* +++<u>have read</u>+++ the _book_ *Graph Databases*.

> You can use this information to build a data model by identifying the components as labels, nodes, and relationships.
> To add it to the graph, you can write the following Cypher statement:

Reviewer: ```suggestion
You can use this information to build a data model by identifying the components as labels, nodes, and relationships.
If you break up the scenario into smaller components, you can define them as parts of your property graph model. 
```

Reviewer: The point here is not to create a Cypher statement, but to build a data model and to understand how the scenario can be broken up and converted into graph concepts.

> [source,cypher]
> --
> CREATE (b:Book {name:'Graph Databases', author:'Jim Webber', author:'Ian Robinson'})<-[:HAS_READ {on:date("2013-03-02"), rated:5}]-(p1:Person {name:'John', age:27})-[:IS_FRIENDS_WITH {since:date("2013-09-01")}]->(p2:Person {name:'Sally', age:32})
> --
> 
> And the visualization should look like this:

Reviewer: ```suggestion
Start by simplifying the model:
```

Reviewer: Again, this section is about understanding your domain and how this can be translated into a data model.

> image::property_graph_elements-arr.svg[700,700,role="popup-link"]

Reviewer: I'd recommend to make a new model and exclude all properties that can't be extracted from the scenario above.

> You can see that:

Reviewer: ```suggestion
In the model, you can see that:
```

* Nodes (circles) represent objects and people.
* Nodes can have properties (name/value pairs).
* Relationships (arrows) connect nodes and represent actions.
* Relationships are directional and can have properties (name/value pairs).

> The domain in this example is, therefore, about people and books, and how they relate to each other.

Reviewer: ```suggestion

```
The domain should already be known, the point here is to use knowledge about the domain to design an appropriate model.
I would also recommend to keep the content about nodes and relationships, but refine it and shorten it.

Author: Yes, I think this line is kind of an overkill in case the reader still didn't understand what is the domain in this case, but alternatively I could move this sentence right after the scenario's description.

Reviewer: The point of this section is not to determine the domain, it should already be known. As mentioned in the beginning of this section, domain-knowledge is important when designing the data model. This section is about using that knowledge. The model should be a reflection of the domain, or _describe_ it, if you will.

=== Refactoring

Data models can be designed in different ways.

> For instance, the same example used previously could have been modeled with John and Sally being two different nodes instead of having their names added as a property for the nodes labeled `Person`.

Reviewer: ```suggestion
In the previous example, the nodes _Sally_ and _John_ both have the same label, `Person` and their names are added as properties.
In this small model, similar results could be achieved by using the names as labels instead.
However, it would be cumbersome if you wanted to add more people to the graph.
In that case, it would be advisable to group all people nodes using the label `Person`.
```

> However, in case more people are added to the graph, it would make more sense to group John and Sally together into bigger categories, like `Person`, to make retrieving data easier.

Reviewer: ```suggestion

```

Updating the way you organize your data (e.g. turning nodes into node properties) is called xref:data-modeling/graph-model-refactoring.adoc[refactoring] and the iterations are called versioning.

=== Design strategy

Neo4j is *schema-free*, which means that the data model can adapt and change easily as a project advances, thus making refactoring and versioning simpler processes.
This allows developers and architects to have more flexibility when determining the structure of the data model and how to define entities for queries.

> Still, since datasets can often share many points in common (e.g. domains, objects, measurements), you can opt to follow strategies that are already known and used in the relevant industry.

Reviewer: ```suggestion
Datasets often have many things in common (for example, domains, objects, measurements), and there are existing strategies you can follow.
```

:page-ad-underline-role: button
:page-ad-underline: Learn more

Data modeling is a practice that defines the logic of queries and the structure of the data in storage.

> It is an important step to take when using any type of database, as it helps ensuring data integrity and improved performance.

Reviewer: ```suggestion
A well-designed model is the key to leveraging the strengths of a graph database as it improves query performance, supports flexible queries, and optimizes storage.
```

== How to create a graph data model

> There are different ways to organize data into a <<data model>>, but most likely that will depend on what questions you want to answer.

Reviewer: ```suggestion
To organize data into a <<data model>>, the first thing to do is to think about what questions you want to answer.
```

> Consider an example case in which you work for a retail company that wants to learn what products customers buy first and what products they buy subsequently.

Reviewer: ```suggestion
For example, assume that you work for a retail company and want to learn what products customers buy first and what products they buy next.
```

To answer that, you need to:

> . Have data on the products sold by the company, who are its customers, and what they purchased at a certain point in time.
> This process is known as "entity extraction".
> . Understand how these entities relate to each other, and what the properties for these entities are (e.g. time of purchase, customer name).
> . Optionally, you can use xref:data-modeling/data-modeling-tools.adoc[no-code data modeling tools] to visualize the model before actually creating it.
> . If satisfied, you can start writing the data into an instance.

Reviewer: ```suggestion
* Have data on the products sold, the customers, and what they purchased at a certain point in time.
This process is known as "entity extraction".
* Understand how these entities relate to each other.
* Think about what other details need to be provided, i.e. what *properties* should be added to these entities (e.g. time of purchase, customer name).
* To visualize the model before you create it, you can use xref:data-modeling/data-modeling-tools.adoc[no-code data modeling tools].
* If satisfied, you can start writing the data into an instance.
```

> In this fictional scenario, you can start by adding this information to the graph:

Reviewer: See comment on MERGE below.

[source,cypher]
--

> MERGE (c:Customer {name: "John"})

Author: Yes, but MERGE also works as CREATE when the data doesn't exist yet. However, I think for didactical reasons, it would be actually better to use CREATE

> MERGE (c:Customer {name: "John"})

Reviewer: Yes, my point exactly!

MERGE (p:Product {name: “Camera”})
MERGE (c)-[:BUYS]->(p)
--

> Then, you can test this model by querying the database with a question (e.g. who bought a camera):

Reviewer: But we said earlier that the question we want to answer is which product a customer buys first and which they buy subsequently. If we ask the model something else, we don't really test the model. If we want a worked example, we should probably include the initial question when we test the model.

Author: Makes sense, but I think it's better to add one more question (which is answered in this test) to justify the part on refactoring.

[source,cypher]
--
MATCH (c)-[b:BUYS]->(p:Product {name:"Camera"})
RETURN c
--

Keep in mind that graph data modeling is an iterative process.

> Your initial graph data model is only a starting point, but as you learn more about the use cases or if the use cases change, the initial graph data model will need to adapt. 
> In addition, you may find that especially when the graph scales, you will need to modify the graph (xref:data-modeling/graph-model-refactoring.adoc[refactor]) to achieve the best performance for your key use cases.

Reviewer: ```suggestion
Your initial graph data model is only a starting point. 
As you learn more about your use cases or if they change, the model needs to adapt.
Additionally, you may find that, especially when the graph scales, you need to xref:data-modeling/graph-model-refactoring.adoc[refactor] your model to ensure it is aligned with your business needs as they evolve.
```

> More generally, you can summarize the process of creating a data model in these steps:

Reviewer: ```suggestion
In summary, the process of creating a data model includes the following:
```

> . Understand the domain and define specific use cases (questions) for the application.
> . Develop the initial graph data model by extracting entities and how they relate to each other.
> . Test the use cases against the initial data model.
> . Create the graph with test data using Cypher.
> . Test the use cases, including performance against the graph.
> . Refactor (improve) the graph data model due to a change in the key use cases or for performance reasons.
> . Implement the refactoring on the graph and retest using Cypher.

Reviewer: ```suggestion
* Understand the domain and define specific use cases (questions) for the application.
* Develop an initial graph data model by extracting entities and decide how they relate to each other.
* Test the use cases against the initial data model.
* Create the graph with test data using Cypher.
* Test the use cases, including performance against the graph.
* Refactor the graph data model due to changes in the key use cases or for performance reasons.
```

Author: I think it's important for this list to be ordered as the topics describe a linear process

Reviewer: That makes sense.

:page-ad-underline-role: button
:page-ad-underline: Learn more

Data modeling is a practice that defines the logic of queries and the structure of the data in storage.
A well-designed model is the key to leveraging the strengths of a graph database as it improves query performance, supports flexible queries, and optimizes storage.

== How to create a graph data model

To organize data into a <<data model>>, the first thing to do is to think about what questions you want to answer.

For example, assume that you work for a retail company and want to learn what products customers are buying.
To answer that:

> * Have data on the products sold, the customers, and what they purchased at a certain point in time.

Reviewer: ```suggestion
* Have data on the products sold and the customers who bought them.
```
If we don't want to know what they bought next, there is no need to know at what time they bought anything.

This process is known as "entity extraction".
* Understand how these entities relate to each other.

> * Think about what other details need to be provided, i.e. what properties should be added to these entities (e.g. time of purchase, customer name).

Reviewer: ```suggestion
* Think about what other details need to be provided, i.e. what properties should be added to these entities (e.g. customer name and product price).
```

* Optionally, visualize the model before you create it using xref:data-modeling/data-modeling-tools.adoc[no-code data modeling tools].
* If satisfied, you can start writing the data into an instance.

In this fictional scenario, you can start by adding this information to the graph:

[source,cypher]
--
CREATE (c:Customer {name: "John"})
CREATE (p:Product {name: “Camera”})
CREATE (c)-[:BUYS]->(p)
--

> Then, you can test this model by querying the database with a question (e.g. what did John buy):

Reviewer: ```suggestion
Then, you can test this model with a query (e.g. what did John buy):
```
A query is essentially the same as a question to a database.
